pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go

```diff
@@ -193,6 +200,226 @@ type HTTPLocalRateLimitSpec struct {
+}
+
+// RateLimitServiceSpec defines the Rate Limit Service specification.
+type RateLimitServiceSpec struct {
+	// Host defines the hostname of the rate limiting service.
+	Host string `json:"host"`
+
+	// Port defines the port number of the rate limiting service
+	Port uint16 `json:"port"`
+}
+
+// TCPRateLimitDescriptor defines the rate limit descriptor to use
+// in the rate limit service request for TCP connections.
+type TCPRateLimitDescriptor struct {
+	// Entries defines the list of rate limit descriptor entries.
+	Entries []TCPRateLimitDescriptorEntry `json:"entries"`
+}
+
+// TCPRateLimitDescriptorEntry defines the rate limit descriptor entry as a
+// key-value pair to use in the rate limit service request for TCP connections.
+type TCPRateLimitDescriptorEntry struct {
+	// Key defines the key of the descriptor entry.
+	Key string `json:"key"`
+
+	// Value defines the value of the descriptor entry.
+	Value string `json:"value"`
+}
+
+// HTTPGlobalRateLimitSpec defines the global rate limiting specification
+// for HTTP requests.
+type HTTPGlobalRateLimitSpec struct {
+	// RateLimitService defines the rate limiting service to use
+	// as a global rate limiter.
+	RateLimitService RateLimitServiceSpec `json:"rateLimitService"`
+
+	// Domain defines a container for a set of rate limits.
+	// All domains known to the Ratelimit service must be globally unique.
+	// They serve as a way to have different rate limit configurations that
+	// don't conflict.
+	Domain string `json:"domain"`
+
+	// Descriptors defines the list of rate limit descriptors to use
+	// in the rate limit service request.
+	Descriptors []HTTPGlobalRateLimitDescriptor `json:"descriptors"`
+
+	// Timeout defines the timeout interval for calls to the rate limit service.
+	// Defaults to 20ms.
+	// +optional
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+
+	// FailureModeDeny defines whether to allow traffic in case of
+	// communication failure between rate limiting service and the proxy.
+	// Defaults to false.
+	// +optional
+	FailureModeDeny *bool `json:"failureModeDeny,omitempty"`
+
+	// EnableXRateLimitHeaders defines whether to include the headers
+	// X-RateLimit-Limit, X-RateLimit-Remaining, and X-RateLimit-Reset on
+	// responses to clients when the rate limit service is consulted for a request.
+	// Defaults to false.
+	// +optional
+	EnableXRateLimitHeaders *bool `json:"enableXRateLimitHeaders,omitempty"`
+}
+
+// HTTPGlobalRateLimitDescriptor defines rate limit descriptor to use
+// in the rate limit service request for HTTP requests.
+type HTTPGlobalRateLimitDescriptor struct {
+	// Entries defines the list of rate limit descriptor entries.
+	Entries []HTTPGlobalRateLimitDescriptorEntry `json:"entries,omitempty"`
+}
+
+// HTTPGlobalRateLimitDescriptorEntry defines the rate limit descriptor entry
+// to use in the rate limit service request for HTTP requests.
+type HTTPGlobalRateLimitDescriptorEntry struct {
```

Comment:
are these intended to act as a `oneof` ?

Reply:
Yes, updated the comment to remove ambiguity.